### PR TITLE
AbstractSampler Interface

### PR DESCRIPTION
**Pull request by tjungni**
TODO:
- [x] Update QEDprocesses in QEDevents on released version
- [x] Update QEDprocesses in QEDevents/test on released version
- [x] Remove qedprocesses_setup_interface.jl moc file
- [x] Replace QEDevents. with QEDprocesses. and remove at QEDevents.compute

### CODE REVIEW

File: src/interfaces/sampler_interface.jl [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+"""
+abstract type AbstractSampler <: AbstractComputationSetup end
+
+@inline Base.eltype(::AbstractSampler) = Float64
+
+"""
+Interface function, which asserts that the given `input` is valid.
+"""
+function _assert_valid_input(smplr::AbstractSampler, x::AbstractVecOrMat)
+    size(x, 1) == size(smplr, 1)[1] || error(
```
Comment (szabo137) [inline src/interfaces/sampler_interface.jl]: Instead of using the function `error("...")`, consider using `throw(InvalidInputError("..."))`.

File: src/interfaces/sampler_interface.jl [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        "Invalid input. The dimensionality of the input is $(size(x,1)) but it should be $(size(smplr,1)[1]).",
+    )
+    return eltype(x) == eltype(smplr) || error(
```
Comment (szabo137) [inline src/interfaces/sampler_interface.jl]: Same: Instead of using the function `error("...")`, consider using `throw(InvalidInputError("..."))`.

File: src/interfaces/sampler_interface.jl [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+function _compute(smplr::AbstractSampler, sample)
+    return _weight(smplr::AbstractSampler, sample)
+end
+
+"""
+
+    _weight(::AbstractSampler, sample)
+
+Return the weight associated with the given sample according to the sampler.
```
Comment (szabo137) [inline src/interfaces/sampler_interface.jl]: The docstring should indicate, that this is an interface function.

Furthermore, the docstring should contain a note, that the implementation of `_weight` must not perform input validation checks (this is done in `weight` via the call of `compute`).

File: src/interfaces/sampler_interface.jl [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+"""
+function _weight end
+
+"""
+
+    setup(::AbstractSampler)
+
+???
```
Comment (szabo137) [inline src/interfaces/sampler_interface.jl]:
```suggestion
Interface function, which returns the `setup::AbstractComputeSetup` associated with the sampler.
```

File: src/interfaces/sampler_interface.jl [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+"""
+function setup end
+
+"""
+
+    is_exact(::AbstractSampler)
+
+Return if the sampler is exactly representing the base distribution or not.
```
Comment (szabo137) [inline src/interfaces/sampler_interface.jl]: The docstring should indicate, that this is an interface function.

File: src/interfaces/sampler_interface.jl [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+"""
+function is_exact end
+
+function weight(smplr::AbstractSampler, sample)
```
Comment (szabo137) [inline src/interfaces/sampler_interface.jl:92]: docstring?

File: src/interfaces/sampler_interface.jl [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    return compute(smplr, sample)
+end
+
+"""
+
+    rand(rng, ::AbstractSampler, ::AbstractVector{P}) where {P}
```
Comment (szabo137) [inline src/interfaces/sampler_interface.jl]:
```suggestion
    _rand!(rng, ::AbstractSampler, ::AbstractVector{P}) where {P}
```

File: src/interfaces/sampler_interface.jl [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+end
+physical_model(smplr::AbstractScatteringProcessSampler) = physical_model(setup(smplr))
+
+@inline Base.eltype(::AbstractScatteringProcessSampler{T}) where {T} = T
+
+"""
+
+    max_weight(::AbstractSampler)
+
+Return the maximum possible weight for the sampler.
```
Comment (szabo137) [inline src/interfaces/sampler_interface.jl]: Same: The docstring should indicate, that this is an interface function.

File: test/interfaces.jl [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+
+struct TestSetup{D} <: QEDevents.AbstractComputationSetup
+    dist::D
+end
+
+function TestSetup(dim::Integer)
+    return TestSetup(product_distribution(Uniform.(-rand(dim), 1.0)))
+end
+
+Base.size(stp::TestSetup, N::Integer) = size(stp.dist)[1] ###best way?
```
Comment (szabo137) [inline test/interfaces.jl]: Better use `size(stp.dist,1)`.

File: test/interfaces.jl [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+@testset "sampler interface" begin
+    setup(smplr::TestSampler) = smplr.stp
+    QEDevents._weight(smplr::TestSampler, x) = QEDevents.compute(smplr.stp, x) # assuming a product of Uniforms    ##remove the QEDevents nstead of replacing by processes because exported
+    max_weight(smplr::TestSampler) = pdf(smplr.stp.dist, zeros(size(smplr.stp.dist)))
+    function Distributions._rand!(
+        rng::Random.AbstractRNG, s::TestSampler, x::AbstractVector{T}
+    ) where {T<:Real}
+        # assuming a sampler setup from Distributions.jl
+        return Distributions.rand!(rng, s.stp.dist, x)
+    end
```
Comment (szabo137) [inline test/interfaces.jl]: This can be placed outside the `@testset`.

File: test/interfaces.jl [hunk 11]
```diff
@@ -0,0 +1,113 @@
+#using QEDprocesses
+using Random
+using Distributions
+using QEDbase
+
+RNG = MersenneTwister(708583836976)
+DIM = rand(RNG, 2:8)
```
Comment (szabo137) [inline test/interfaces.jl]: Maybe also cover the edge-case `DIM=1`.

File: test/interfaces.jl [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+    end
+
+    @testset "process sampler interface" begin
+        proc_stp = TestSetup(DIM)
+        test_smplr = TestSampler(proc_stp)
+
+        @testset "properties" begin
+            @test setup(test_smplr) == proc_stp
+            @test is_exact(test_smplr) == RAND_EXACTNESS
+            @test size(test_smplr) == size(proc_stp)
```
Comment (szabo137) [inline test/interfaces.jl]: `DIM?`

File: test/interfaces.jl [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+        end
+
+        @testset "weight: vector" begin
+            x_out = rand(RNG, DIM)
+            test_vals = weight(test_smplr, x_out)
+            groundtruth = pdf(proc_stp.dist, x_out)
+            @test isapprox(test_vals, groundtruth, atol=ATOL, rtol=RTOL)
+        end
+
+        @testset "weight: vector-matrix" begin
```
Comment (szabo137) [inline test/interfaces.jl]:
```suggestion
        @testset "weight: matrix" begin
```

File: test/interfaces.jl [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+            # reproduce the same random samples three times
+            rng1 = deepcopy(RNG)
+            rng2 = deepcopy(RNG)
+            rng3 = deepcopy(RNG)
+
+            test_x_inplace = zeros(DIM)
+            rand!(rng1, test_smplr, test_x_inplace)
+            test_x = rand(rng2, test_smplr)
+            groundtruth = rand(rng3, proc_stp.dist)
+            ###_____versions
```
Comment (szabo137) [inline test/interfaces.jl]: ??

File: test/runtests.jl [hunk 15]
```diff
@@ -2,5 +2,5 @@ using QEDevents
 using Test
 
 @testset "QEDevents.jl" begin
-    # Write your tests here.
+    include("interfaces.jl")
```
Comment (szabo137) [inline test/runtests.jl]: Consider using `@safetestsets` from `SafeTestsets`. For instance, see https://github.com/QEDjl-project/QEDprocesses.jl/blob/dev/test/runtests.jl

**Review by szabo137 — COMMENTED** [review]
First comments and questions.

File: test/interfaces.jl (hunk 9 above)
Comment (tjungni) [reply test/interfaces.jl]: That doesn't work:
```julia
MethodError: no method matching size(::Product{Continuous, Uniform{Float64}, Vector{Uniform{Float64}}}, ::Int64)
```

File: src/interfaces/sampler_interface.jl (hunk 1 above)
Comment (tjungni) [reply src/interfaces/sampler_interface.jl]: Implemented in 29e04d4dc84862ce3aabe9a8addce347bdc5f0a7

File: src/interfaces/sampler_interface.jl (hunk 2 above)
Comment (tjungni) [reply src/interfaces/sampler_interface.jl]: Implemented in 29e04d4dc84862ce3aabe9a8addce347bdc5f0a7

File: test/interfaces.jl (hunk 10 above)
Comment (tjungni) [reply test/interfaces.jl]: Implemented in 40475f80c9b50a7d4831a5316723372b2a6c56a9

File: test/interfaces.jl (hunk 12 above)
Comment (tjungni) [reply test/interfaces.jl]: Implemented in 40475f80c9b50a7d4831a5316723372b2a6c56a9

File: test/interfaces.jl (hunk 14 above)
Comment (tjungni) [reply test/interfaces.jl]: Removed in 40475f80c9b50a7d4831a5316723372b2a6c56a9

File: test/runtests.jl (hunk 15 above)
Comment (tjungni) [reply test/runtests.jl]: Implemented in 40475f80c9b50a7d4831a5316723372b2a6c56a9

File: src/interfaces/sampler_interface.jl (hunk 3 above)
Comment (tjungni) [reply src/interfaces/sampler_interface.jl]: Added in e91af7c76038a6cbcaeb764333ea07da0e5bacab

File: src/interfaces/sampler_interface.jl (hunk 5 above)
Comment (tjungni) [reply src/interfaces/sampler_interface.jl]: Added in e91af7c76038a6cbcaeb764333ea07da0e5bacab

File: src/interfaces/sampler_interface.jl (hunk 6 above)
Comment (tjungni) [reply src/interfaces/sampler_interface.jl:92]: Added in e91af7c76038a6cbcaeb764333ea07da0e5bacab

File: src/interfaces/sampler_interface.jl (hunk 8 above)
Comment (tjungni) [reply src/interfaces/sampler_interface.jl]: Added in e91af7c76038a6cbcaeb764333ea07da0e5bacab

File: test/interfaces.jl (hunk 11 above)
Comment (tjungni) [reply test/interfaces.jl]: Added in c8a7e0dad3a11dcfd62337ab780529620313383f

File: src/interfaces/sampler_interface.jl [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+
+@inline Base.eltype(::AbstractSampler) = Float64
+
+"""
+Interface function, which asserts that the given `input` is valid.
+"""
+function _assert_valid_input(smplr::AbstractSampler, x::AbstractVecOrMat)
+    size(x, 1) == size(smplr, 1)[1] || throw(
+        InvalidInputError(
+            "Invalid input. The dimensionality of the input is $(size(x,1)) but it should be $(size(smplr,1)[1]).",
```
Comment (szabo137) [inline src/interfaces/sampler_interface.jl]:
```suggestion
            "The dimensionality of the input is $(size(x,1)) but it should be $(size(smplr,1)[1]).",
```

File: src/interfaces/sampler_interface.jl [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+        ),
+    )
+    return eltype(x) == eltype(smplr) || throw(
+        InvalidInputError(
+            "Invalid input. The element type of the input is $(eltype(x)) but it should be $(eltype(smplr)).",
```
Comment (szabo137) [inline src/interfaces/sampler_interface.jl]:
```suggestion
            "The element type of the input is $(eltype(x)) but it should be $(eltype(smplr)).",
```

File: src/interfaces/sampler_interface.jl [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+
+"""
+
+    _weight(::AbstractSampler, sample)
+
+Interface function, which returns the weight associated with the given sample according to the sampler.
+
+!!! note ""
+    
+    This function must not do input validation. This is done by [`weight`](@ref) which calls _weight after input validation.
```
Comment (szabo137) [inline src/interfaces/sampler_interface.jl]:
```suggestion
    This function must not do input validation. This is done by [`weight`](@ref) which calls `_weight` after input validation.
```

File: test/interfaces.jl (hunk 9 above)
Comment (szabo137) [reply test/interfaces.jl]: Ah ok, this is the weird implementation of `size` in `Distributions.jl`. Then ignore my comment.

File: test/interfaces.jl [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+using Random
+using Distributions
+using QEDbase
+
+RNG = MersenneTwister(708583836976)
+
+ATOL = 0.0
+RTOL = sqrt(eps())
+
+RAND_MAX = rand(RNG)
```
Comment (szabo137) [inline test/interfaces.jl]:
```suggestion

```

This variable is never used.

File: test/runtests.jl [hunk 20]
```diff
@@ -1,6 +1,9 @@
 using QEDevents
 using Test
+using SafeTestsets
 
 @testset "QEDevents.jl" begin
-    # Write your tests here.
+    @time @safetestset "interfaces" begin
```
Comment (szabo137) [inline test/runtests.jl]:
```suggestion
    @time @safetestset "sampler interfaces" begin
```

**Review by szabo137 — COMMENTED** [review]
More comments. Furthermore, consider renaming `test/interfaces.jl` to `test/interfaces/sampler_interface.jl`. There will be more interfaces to be tested.

File: src/interfaces/sampler_interface.jl [hunk 21]
```diff
@@ -0,0 +1,140 @@
...
+    return compute(smplr, sample)
+end
+
+"""
+
+    _rand!(rng, ::AbstractSampler, ::AbstractVector{P}) where {P}
+
+Generate a random sample from the sampler and write it into the given vector.
+"""
+function _rand! end
```
Comment (szabo137) [inline src/interfaces/sampler_interface.jl]: This is a type of piracy, since `Random._rand!` and `Distributions._rand!` are already implemented. I suggest removing this and dispatching one of the above (presumably from `Distributions`).

At some point, one could consider to directly interface with `Distributions.jl` via `Sampleable` and related functions (e.g. see https://github.com/JuliaStats/Distributions.jl/blob/e407fa5fd098e50df51801c6d062946eac7a7d0f/src/common.jl#L72).

**Review by szabo137 — CHANGES_REQUESTED** [review]
In general, the PR looks good to me, and after resolving the `_rand!` piracy, it can be merged. 

@tjungni could you please update the bullet points in the description?

File: src/interfaces/sampler_interface.jl (hunk 21 above)
Comment (tjungni) [reply src/interfaces/sampler_interface.jl]: I removed the line in 2ad8208d75d466008955007b234b7999173f2368 so now it should be using Random._rand! since that was already imported in QEDevents.jl

Comment (tjungni) [reply src/interfaces/sampler_interface.jl]: As of d4b5bb079f73aff9a8c1224c9dbdbfc67514f393 all variations of rand are imported from Distributions.jl

File: src/interfaces/sampler_interface.jl [hunk 22]
```diff
@@ -0,0 +1,139 @@
...
+    return compute(smplr, sample)
+end
+
+"""
+
+    _rand!(rng, ::AbstractSampler, ::AbstractVector{P}) where {P}
+
+Generate a random sample from the sampler and write it into the given vector.
+"""
+
```
Comment (szabo137) [inline src/interfaces/sampler_interface.jl:105]:
```suggestion
function _rand!(rng::AbstractRNG, smplr::AbstractSampler, x::AbstractVector)
    throw(MethodError(_rand!,(rng,smplr,x)))
end
```

Comment (szabo137) [reply src/interfaces/sampler_interface.jl:105]: One needs to add a symbol here, because the docstring would not be shown somewhere.

**Review by szabo137 — COMMENTED** [review]
One last comment, then I am happy to merge this!

**Review by szabo137 — DISMISSED** [review]
Thanks for providing this PR, it looks fine to me now and is ready for merging.

@tjungni Could you please run the formatter on the last version of your contribution?

File: src/interfaces/sampler_interface.jl [hunk 23]
```diff
@@ -0,0 +1,142 @@
...
+    optional
+
+    ```Julia
+    _post_processing
+    ```
+
+"""
+abstract type AbstractSampler <: AbstractComputationSetup end
+
+@inline Base.eltype(::AbstractSampler) = Float64
```
Comment (AntonReinhard) [inline src/interfaces/sampler_interface.jl:35]: Is this implementation useful for anything? Wouldn't it lead to someone implementing a `CustomSampler`, forgetting to implement this, but then code still working but with a wrong element type?

File: src/interfaces/sampler_interface.jl [hunk 24]
```diff
@@ -0,0 +1,142 @@
...
+
+"""
+
+    _weight(::AbstractSampler, sample)
+
+Interface function, which returns the weight associated with the given sample according to the sampler.
+
+!!! note ""
+    
+    This function must not do input validation. This is done by [`weight`](@ref) which calls `_weight` after input validation.
```
Comment (AntonReinhard) [inline src/interfaces/sampler_interface.jl]:
```suggestion
    This function must not do input validation. This is done by [`weight`](@ref), which calls `_weight` after input validation.
```

File: src/interfaces/sampler_interface.jl [hunk 25]
```diff
@@ -0,0 +1,142 @@
...
+
+Interface function, which returns whether the sampler is exactly representing the base distribution or not.
+"""
+function is_exact end
+
+"""   
+
+    $(TYPEDSIGNATURES)
+
+Interface function, which validates the input, calculates the weight via [`_weight`](@ref) and performs an optional post-processing via QEDprocesses._post_processing.
```
Comment (AntonReinhard) [inline src/interfaces/sampler_interface.jl]:
```suggestion
Interface function, which validates the input, calculates the weight via [`_weight`](@ref), and performs an optional post-processing via `QEDprocesses._post_processing`.
```

File: test/interfaces/sampler_interface.jl [hunk 26]
```diff
@@ -0,0 +1,116 @@
+using QEDprocesses
+using QEDevents
+using QEDbase
+import Random: AbstractRNG, MersenneTwister
+using Distributions
+import Distributions: rand, rand!, _rand!
+
+RNG = MersenneTwister(708583836976)
+
+ATOL = 0.0
```
Comment (AntonReinhard) [inline test/interfaces/sampler_interface.jl:10]: atol is zero by default anyways and should pretty much never be not 0 unless we're expecting a zero in what we're `isapprox`ing, I think this can go.

**Review by AntonReinhard — CHANGES_REQUESTED** [review]
Is there any specific reason why the parameters are called `smplr` instead of `sampler` everywhere? Am I not seeing a name shadowing? Otherwise, it seems a little obtuse for 2 saved characters.

File: src/interfaces/sampler_interface.jl (hunk 23 above)
Comment (szabo137) [reply src/interfaces/sampler_interface.jl:35]: This is necessary to correctly allocate Vectors for results of a sampler, e.g. as seen here (cf line 118ff): 

```julia
function rand(rng::AbstractRNG, smplr::AbstractSampler)
    return _rand!(rng, smplr, Vector{eltype(smplr)}(undef, size(smplr, 1)))
end
```

File: test/interfaces/sampler_interface.jl (hunk 26 above)
Comment (szabo137) [reply test/interfaces/sampler_interface.jl:10]: Correct, indeed. However, this explicit setting of `atol` makes your point independent from the default behavior of `isapprox`, which could change in future releases of Julia.

**Comment by szabo137** [comment]
@AntonReinhard I replyed to all of your comments. Please approve, if you are happy with this 😺

**Comment by tjungni** [comment]
> @AntonReinhard I replyed to all of your comments. Please approve, if you are happy with this 😺

Thank you!

File: src/interfaces/sampler_interface.jl (hunk 23 above)
Comment (AntonReinhard) [reply src/interfaces/sampler_interface.jl:35]: I wasn't questioning whether the function is needed at all, only whether this default implementation of it is needed for AbstractSamplers

File: test/interfaces/sampler_interface.jl (hunk 26 above)
Comment (AntonReinhard) [reply test/interfaces/sampler_interface.jl:10]: That's fair enough, though that would be a very strange decision ;)

**Review by AntonReinhard — APPROVED** [review]

File: src/interfaces/sampler_interface.jl (hunk 23 above)
Comment (szabo137) [reply src/interfaces/sampler_interface.jl:35]: Ah, ok, understood. I think this is a design choice, whether an `AbstractSampler` returns a vector of floats or a vector of four-momenta, and we thought, the former is more convenient.
